From: Ben Hutchings <ben@decadent.org.uk>
Subject: [PATCH 04/23] bna: fix stats handling

Fixes the same bugs as
commit 250e061e1d3e26600236a3dd9172e7f5f5916c00 upstream, from which the
following description is taken:

get_stats() method incorrectly clears a global array before folding
various stats. This can break SNMP applications.

Signed-off-by: Ben Hutchings <ben@decadent.org.uk>
---
--- a/drivers/net/bna/bnad.c
+++ b/drivers/net/bna/bnad.c
@@ -2035,7 +2035,7 @@ bnad_netdev_qstats_fill(struct bnad *bnad)
 		((u64)bnad->stats.bna_stats->rxf_bmap[1] << 32);
 	for (i = 0; bmap && (i < BFI_LL_RXF_ID_MAX); i++) {
 		if (bmap & 1) {
-			net_stats->rx_fifo_errors =
+			net_stats->rx_fifo_errors +=
 				bnad->stats.bna_stats->
 					hw_stats->rxf_stats[i].frame_drops;
 			break;
@@ -2643,17 +2643,22 @@ bnad_start_xmit(struct sk_buff *skb, struct net_device *netdev)
 {
 	struct bnad *bnad = netdev_priv(netdev);
 	unsigned long flags;
+	unsigned i;
 
 	spin_lock_irqsave(&bnad->bna_lock, flags);
 
 	memset(&bnad->net_stats, 0, sizeof(struct net_device_stats));
 
 	bnad_netdev_qstats_fill(bnad);
 	bnad_netdev_hwstats_fill(bnad);
+
+	for (i = 0; i < sizeof(struct net_device_stats) / sizeof(unsigned long); i++)
+		((unsigned long *)&netdev->stats)[i] =
+			((unsigned long *)&bnad->net_stats)[i];
 
 	spin_unlock_irqrestore(&bnad->bna_lock, flags);
 
-	return &bnad->net_stats;
+	return &netdev->stats;
 }
 
 static void
